### DISCUSSION TITLE
Water Tank Monitoring System on Solar Power

### DISCUSSION THREAD

**Discussion by vic42** [arendst/Tasmota#13354] [Show and tell]
# Solar Powered Tank Monitoring System
![Spectacle s23268](https://user-images.githubusercontent.com/3363362/137123617-6fcd6876-d39a-41a0-9233-f807e328dd1f.png)
## Purpose
The system was created to monitor the water level of a cistern that supplies fresh water from a well to the house and, to communicate through MQTT to the well pump & valve controller in order to maintain a desired water level.
## Design
The cistern's access hatch is covered by a steel lid that has a 20 mm dia hole at the center. The basic design idea was to place an autonomous waterproof box on top of the lid and shoot a laser through a window at its bottom to the water surface, in order to measure the water level. I decided to use a Wemos board and chinese modules because I had only very basic soldering equipment available on-site.
### Dimensions of the tank:
- capacity 5 m³
- service height 1000 mm
- access hatch height 600 mm
### Materials I bought from the local hardware store
- the largest weatherproof cable box they had on offer (IP44)
- flat and L-shaped aluminum profiles
- assorted stainless M4 screws, nuts, and washers (A2)
- perforated mounting strip
- foam tape
- a tube of silicone sealant
### The electronics I had brought from my lab for the project
- WEMOS D1 Mini
![Spectacle u23268](https://user-images.githubusercontent.com/3363362/137132893-4201309f-4134-48e2-99a9-fcec9912a2bf.png)
- VL53L1X breakout board with waterproof window
![Spectacle x23268](https://user-images.githubusercontent.com/3363362/137132505-9d7848a3-c11c-4d8e-9308-3ba67beaa2af.png)
- 100 kΩ resistor
- Dupont cables
- 6V solar panel (https://www.aliexpress.com/item/1005002655207675.html)
![Spectacle V23268](https://user-images.githubusercontent.com/3363362/137122994-b933b7bb-e6f9-427a-b274-8210c083b268.png)
- 10 Ah Li-ion battery (https://www.aliexpress.com/item/1005002257847984.html)
![Spectacle E23268](https://user-images.githubusercontent.com/3363362/137121578-cdf9a741-f100-47f7-a620-78f02928b929.png)
 - 1S solar charger board (https://www.aliexpress.com/item/32862470804.html)
![Spectacle H23268](https://user-images.githubusercontent.com/3363362/137120830-c38ea571-464d-4d26-a5f5-22c65baa10c4.png)
- 5 V / 1 A power bank module (HP-0014B PRO)
![Spectacle F23268](https://user-images.githubusercontent.com/3363362/137130174-5e212547-ffa3-4f9b-838d-d3604e01d1c6.png)
- 3.3V switching buck converter
![Spectacle A23268](https://user-images.githubusercontent.com/3363362/137131913-cdad06df-e5e7-4850-85b1-7602e3e2ff31.png)
### Upgrade
- BME280 I2C environmental monitoring
<img width="200" alt="Spectacle c21874" src="https://user-images.githubusercontent.com/3363362/145400380-3c5cf9e4-e898-45c7-a95f-80b73f1a4eea.png">

## Build
- cut, bent and drilled the aluminum profiles to form a frame to hold the solar panel at about the ideal inclination at winter time for the deployment area. Screwed the frame onto the lid of the box, generously sealed with silicone.
![Spectacle r23268](https://user-images.githubusercontent.com/3363362/137143863-ee70884b-334e-4427-8797-7486c4e0e87a.png)
- drilled an oval hole into the bottom of the box for the VL53L1X window to fit into exactly, carefully applied some silicone sealant on its body and hot-glued the sensor into place.
![Spectacle e23268](https://user-images.githubusercontent.com/3363362/137133651-4391333e-be6e-49de-8cb3-72a158620076.png)
- applied foam tape to the bottom of the box and the mounting strip and placed the battery in between. Screwed down the mounting strip to hold the battery in place.
- connected and hot-glued all other PCB modules in place
- glued two headers to the side wall to plug in the Wemos in order to keep it removable.
- added a 100k resistor into the yellow cable in order to adjust the Wemos ADC input to 4.2V max to measure the battery voltage. Calibrated the battery readings against my multimeter
![Spectacle H23268](https://user-images.githubusercontent.com/3363362/137137949-4708ea81-2483-4b31-8f4f-092454e72445.png)
- drilled four holes into the metal sheet to firmly mount the box onto the removable tank lid.
## Operation
I was prepared to send the ESP into deep sleep to save power but surprisingly the dimensioning of the solar system never allowed the battery voltage to drop below 4 Volts so far, while the ESP is still running at full steam with a `TelePeriod` of 60 seconds.
![Spectacle l23268](https://user-images.githubusercontent.com/3363362/137138961-a193ae9a-637a-422d-ac11-58ef8af51a94.png)
The diagram shows a week's worth of battery voltage readings. 
## Conclusion
This device works surprisingly well. I think this design could serve as a template for many outdoor sensing applications. Being the first of its kind, it is built like a mash-up, proof of concept thing, built from the modules randomly at hand. The next one will definitely receive updates to the circuitry as part of a clean design. It remains to be seen how the system will survive the particularly rainy winter time at southern Greece.

**Comment by sfromis** [comment]
Since IP44 is "only" splash-proof, I'm wondering if you do something to manage or measure humidity inside the box. Seasonal and daily variations might lead to condensation under some circumstances.

**Reply by vic42** [reply]
I inspected the box and found it to be actually air tight as long as you don't use the cable openings. It has a very good "sausage" style seal under the hood. So I bear some hope that the inside will remain dry. I will definitely do an inspection after its first winter.

**Reply by sfromis** [reply]
Fair enough. My curiosity would probably still have inspired me to have a BME280 or similar inside :-D

**Reply by vic42** [reply]
I can add a BME280 next time to satisfy your curiosity :-)

**Reply by barbudor** [reply]
Nice build @vic42 👍
Regarding humidity, I think the risk is more with condensation than humidity in the air. Of course condensation would probably not occur if there is not humidity in the air first.

**Reply by sfromis** [reply]
Yeah, as long as the seal seems to be much better than IP44, low risk of humid air ingress with pressure variations.

**Reply by techman83** [reply]
Probably worth throwing a couple of desiccant packets in there. Water molecules are tiny, so sealing things entirely can be a bit of a losing battle. But you can get away with a lot. I have a cheap Temp/Humidity sensor just hanging out of my roof with a bare PCB. Figured I'd replace it and try something else if it failed too quickly. Lasted around 7 years before it failed, so I just re-terminated the wires (had left plenty of slack) and plugged in a new one!

**Reply by barbudor** [reply]
Looking forward to see how the battery & solar charging evolves in time.
I set up my ES8266 based outdoor sensor in summer 2020 with a 1700mAh battery and a 0.5W Solar panel (yes 10 times lower than yours).

I was not able to keep it running without DeepSleep after September.
1 little more than 1 year later, sun is not enough to charge the battery anymore.
If I fully charge the battery from an adapter, even with DeepSleepTime of 30 minutes, I can't get more than 2 days.
I believe the battery capacity has drastically fallen due to high temperature changes across the year. 

You definitively chose a stronger capacity for your solar panel and your battery.

**Reply by electrofun-smart** [reply]
Great project! Can you share the electronic schematic?

**Reply by electrofun-smart** [reply]
Do you think the ultrasonic sensor SR04 can be used instead of laser sensor VL53L1X? What are the pros and cons?

**Reply by sfromis** [reply]
The cons is that it is a worse sensor, less stable, and not good for water-proofing. Basically no pros for this.

**Reply by electrofun-smart** [reply]
> The cons is that it is a worse sensor, less stable, and not good for water-proofing. Basically no pros for this.

Thanks! I'm planning to make a similar project, but the area where the water tank is located has no wifi connection, the distance is around 100 meters from the house. I saw that I can either use LORA, ESP-NOW or RF433MhZ, do you have any suggestion what is better?

**Reply by sfromis** [reply]
You "saw"? Many technologies exists, but I'm not sure that Tasmota has much support outside of what it was designed for, which includes a good power source, and wifi. 

ESP-NOW is basically wifi, same radios.

**Reply by vic42** [reply]
> Since IP44 is "only" splash-proof, I'm wondering if you do something to manage or measure humidity inside the box. Seasonal and daily variations might lead to condensation under some circumstances.

I just followed your suggestion and added a BME280 to the box. Now we can monitor environmental conditions and see if there is a humidity problem.
<img width="277" alt="vl53bme280" src="https://user-images.githubusercontent.com/3363362/145399674-2b45763f-96e3-48ca-8b54-e45e3451fd78.png">

**Reply by vic42** [reply]
Unfortunately, I had to learn that the BME280 probably needs some air flow in order to produce sensible humidity readings. Once you close the lid of the box humidity maxes out at 100% very quickly although the box is completely dry inside. The other readings are stll worthwile, however.

**Reply by sfromis** [reply]
I actually did a small test with a BME280 in a closed jar, without airflow. Nothing weird about humidity readings, just kept the same quite plausible level.

**Reply by barbudor** [reply]
It may takes time...
But I also have a problem with my outdoor sensor where the BME280 is outside of the box, protected from direct rain but it should have some air flow. However the humidity is now continuously at 100% since months.
I would assume that some condensation occured in the sensor itself.

**Reply by sfromis** [reply]
Your sensor may be busted. What vic posted about (on Discord) was it very quickly rising to 100% when the box was closed. My test was for nearly 2 days.

My current outdoors BME280 (in a plastic enclosure with slits for ventilation) gives plausible numbers, currently mostly above 90% (I'm not expecting great accuracy at high humidity). While it sometimes flatlines at 100% for hours, the surrounding plot points does not make it look like it was stuck, merely maxed out, possibly not really able to distinguish differences at the high humidity level.

My previous outdoors BME280 was housed in a way less reliably protecting it from becoming wet. It failed (producing too high temperatures), and when inspecting, I found that one pin was so corroded that it fell off. Currently the replacement is paired with a BMP280, as I assume that both would not fail the same way :-)

**Comment by lwyl5744** [comment]
very interesting is there a wiring diagram and code available

**Reply by vic42** [reply]
There is not much code involved, just Tasmota. (And the typical node-red backend plus some php web graphics using RRDTool)

The wiring is so obvious that I did not even draw a diagram.

- The sensors are both i2c, SCL and SDA wired to GPIO4, GPIO5 of the Wemos as usual.
- Battery monitoring is done by connecting B+ through a 100k resistor in series into A0 of the Wemos.
- The solar charger connects to the panel on one side and the battery on the other side.
- DC/DC-Converters: Battery -> 5V boost -> 3.3V buck -> Wemos (Not using the linear regulator on the Wemos board due to bad efficiency)

That's pretty much it.
If you still can't figure out how to wire this up I may add a drawing.

**Comment by Yousefff1** [comment]
Hi @vic42,

I have a similar setup, I'm facing an issue though, it's really warm where I live, and I started noticing short readings between specific hours, I checked the sensor, and there seems to be condensation on the lens of the Vl53l1x, I assume because of water vapour that rises and hits the tank lid, and that's where the sensor is placed.
Did you face such issue? Do you know how to get around that?

**Reply by sfromis** [reply]
You should avoid the sensor as such getting wet, possibly by enclosing it to avoid condensation directly on the breakout board. If enclosed, one technique to reduce condensation effects would be the same as for a car window, as adding a bit of soap tends to avoid condensation droplets. Of course, heating or a wiper is also possible, but may not be feasible.

**Reply by Yousefff1** [reply]
@sfromis thank you for the reply, the pcb and the sensor are all waterproof, I've designed a custom water-tight enclosure and used a 3D printer to make it, and all the gaps are covered with silicon.
But for 2-3 hours a day there seems to be water droplets on the sensor lens (external), and that blocks the readings (or distorts them).
Unfortunately I can't use soap, as this tank is a fresh water tank, and I don't want to contaminate it.

**Reply by vic42** [reply]
Hi Youssefff1,

I do get these odd readings once in a while. I assume reflections from mist in the tank. Currently I'm filtering in software but I'm thinking about modifying the driver in order to reduce sensitivity of the sensor to focus on the strongest echo. I currently lack the time to work on it.

![tlevel_d](https://user-images.githubusercontent.com/3363362/161996218-d83b322f-c035-49b3-95ef-2aaa51f6efc9.png)

I used an ultrasonic transducer (sealed and water tight) for another tank that doesn't exhibit this behavior. But the downside of that one is a dead zone of ~300 mm (typical for single transducer sensors).

**Reply by Yousefff1** [reply]
Thank you @vic42,
In my case, those readings are constant for 2-4 hours, I assume water dries out after that.
Here are the readings from today:
![Screen Shot 2022-04-06 at 6 47 21 PM](https://user-images.githubusercontent.com/50089643/162015492-13270736-023b-4ac4-8914-f70715649bcf.png)

**Reply by Yousefff1** [reply]
@vic42 what's interesting is that we are both seeing the same effect (around the same time of the day), which I assume is when the sensor is cold and the water is getting warmer, that supports condensation being the main culprit and not water mist, otherwise we would see false readings throughout the day, and more in the afternoon vs only for a few hours in the morning.

I assume the ultrasonic transducer didn't exhibit such effects because of the large sensor detection area, I doubt any single droplet will block the entire reception area at any given time.

I'm considering a physical cover that opens just before taking the reading, but this comes with it's own complexities.

Btw did you see this effect in colder times? Since you had the sensor installed before I did.

**Comment by barbudor** [comment]
Ventilation could be a solution which also comes with its own complexities
